### PR TITLE
GeoIP banning from ipdeny list

### PR DESCRIPTION
**Pull request by stephdl**
Add geo IP banning from ipdeny website, once downloaded the List is updated each month with a random time

### CODE REVIEW

File: root/etc/e-smith/db/configuration/defaults/blacklist/geoipStatus [hunk 1]
```diff
@@ -0,0 +1 @@
+disabled
```
Comment (gsanchietti) [inline root/etc/e-smith/db/configuration/defaults/blacklist/geoipStatus:1]: The first letter of the prop should be in upper case to mach all other props.
Please change it to `GeoipStatus`

File: api/ipsets/update [hunk 2]
```diff
@@ -26,14 +26,15 @@
 data=$(cat /dev/stdin)
 
 status=$(echo $data | jq -r '.status')
+geoipStatus=$(echo $data | jq -r '.geoipStatus')
```
Comment (gsanchietti) [inline api/ipsets/update:29]: The first letter of the prop should be in upper case to mach all other props.
Please change it to `GeoipStatus`.

As a reference, I will going to propose fixes only of this file but please replace all occurrences.

```suggestion
geoipStatus=$(echo $data | jq -r '.GeoipStatus')
```

File: api/ipsets/update [hunk 3]
```diff
@@ -26,14 +26,15 @@
...
 url=$(echo $data | jq -r '.Url')
 whitelist=$(echo $data | jq -r '.Whitelist | join(",")')
 categories=$(echo $data | jq -r '.Categories | join(",")')
 
 current_url=$(/sbin/e-smith/config getprop blacklist Url)
 
 if [ "$status" = "enabled" ]; then
-    /sbin/e-smith/config setprop blacklist status "$status" Url "$url" Whitelist "$whitelist" Categories "$categories"
+    /sbin/e-smith/config setprop blacklist status "$status" geoipStatus "$geoipStatus" Url "$url" Whitelist "$whitelist" Categories "$categories"
```
Comment (gsanchietti) [inline api/ipsets/update:37]:
```suggestion
    /sbin/e-smith/config setprop blacklist status "$status" GeoipStatus "$geoipStatus" Url "$url" Whitelist "$whitelist" Categories "$categories"
```

File: root/etc/e-smith/events/actions/nethserver-blacklist-conf [hunk 4]
```diff
@@ -27,15 +27,30 @@ event=$1    # name of the event (not used)
...
+    ipsetUrl=$(/sbin/e-smith/config getprop blacklist Url)
+    geoipStatus=$(/sbin/e-smith/config getprop blacklist geoipStatus)
+
+    if [[ "$status" != "enabled" ]]; then
+        exit 0
+    fi
+
+    if [[ "$ipsetUrl" != "" ]]; then
+        /usr/share/nethserver-blacklist/download ipsets
+    fi
```
Comment (gsanchietti) [inline root/etc/e-smith/events/actions/nethserver-blacklist-conf:40]: Is this a real condition?

File: root/etc/e-smith/templates/etc/cron.d/nethserver-blacklist/10base [hunk 5]
```diff
@@ -14,4 +17,12 @@
     } else {
         $OUT .= "# Blacklists (dns) are disabled"
     }
+
+    $OUT .= "\n";
+
+    if (($ip_status eq 'enabled') && ($geoipStatus eq 'enabled')) {
+        $OUT .= '* * 1 * * root sleep $(( ( RANDOM \% 7200 ) )); /usr/share/nethserver-blacklist/download geoip'
```
Comment (gsanchietti) [inline root/etc/e-smith/templates/etc/cron.d/nethserver-blacklist/10base:24]: GeoIP lists are updated quite often, so I would download it once a day.

File: root/usr/share/nethserver-blacklist/geoip [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    echo "#
+# ipv4 hash:ip ipset
+# Category : geoip_blocking
+# Source File Date : $(/usr/bin/date)
+# Maintainer : NethServer
+# List source URL : https://www.ipdeny.com/ipblocks/data/countries/all-zones.tar.gz
+# ISO Country Code : ${j%.zone}
+#
+$(/usr/bin/cat ${j})" > ${j}
+    /usr/bin/mv -- "$j" "ISO_country_code_${j%.zone}.netset"
```
Comment (gsanchietti) [inline root/usr/share/nethserver-blacklist/geoip:38]: Why not creating files directly to the destination dir and avoid the `mv` on line 43?

Also make sure to empty the destination dir before writing down new files.

**Review by gsanchietti — CHANGES_REQUESTED** [review]

File: api/ipsets/validate [hunk 7]
```diff
@@ -63,12 +64,21 @@ else:
...
+        geoipStatus = input_json[geoipStatus_p]
+
+        if geoipStatus not in ['enabled', 'disabled']:
+            invalid_attributes.append(invalid_attribute(geoipStatus_p, "invalid"))
     # url
-    if (url_p not in input_json) or (not input_json[url_p]):
+    if (url_p not in input_json):
         invalid_attributes.append(invalid_attribute(url_p, "empty"))
-    else:
+    elif (input_json[url_p]):
```
Comment (andre8244) [inline api/ipsets/validate:81]: This change makes me think that URL is not mandatory anymore, but [inside download script](https://github.com/NethServer/nethserver-blacklist/blob/1353ae82bb0fbb2309aff196aac9b4eeee274f49/root/usr/share/nethserver-blacklist/download#L81) it looks like it is still mandatory. Am i missing something? :thinking:

**Review by andre8244 — COMMENTED** [review]
UI code looks good to me. I just have some doubts about URL validation

File: root/etc/e-smith/db/configuration/defaults/blacklist/geoipStatus (hunk 1 above)
Comment (stephdl) [reply root/etc/e-smith/db/configuration/defaults/blacklist/geoipStatus]: done

File: api/ipsets/update (hunk 2 above)
Comment (stephdl) [reply api/ipsets/update]: done

File: api/ipsets/update (hunk 3 above)
Comment (stephdl) [reply api/ipsets/update]: done

File: root/etc/e-smith/templates/etc/cron.d/nethserver-blacklist/10base (hunk 5 above)
Comment (stephdl) [reply root/etc/e-smith/templates/etc/cron.d/nethserver-blacklist/10base]: done

File: root/usr/share/nethserver-blacklist/geoip (hunk 6 above)
Comment (stephdl) [reply root/usr/share/nethserver-blacklist/geoip:38]: empty folder before to move is ensured now, I would prefer to use a mktemp

File: api/ipsets/validate (hunk 7 above)
Comment (stephdl) [reply api/ipsets/validate:81]: not a python guru like you @andre8244, I tested some ways. 

Now the url is not a mandatory, if the URL exists then we download (see https://github.com/NethServer/nethserver-blacklist/pull/37/files#diff-af724284fab7ff3dbe0e6c5d083dd81c6c30bddb4a030f75431ef294796b3404R38) same in the cron job ( see https://github.com/NethServer/nethserver-blacklist/pull/37/files#diff-78e4181675f2cbc15bc508d9a79c151096229bc2a3a4cb4bff0a29957c88b9e4R4)

Another way could be to have a third checkbox in the UI to enable git blacklist and a new property in the esmith key

we still check if the url is a valid url and if it is a git repository

File: root/etc/e-smith/events/actions/nethserver-blacklist-conf (hunk 4 above)
Comment (stephdl) [reply root/etc/e-smith/events/actions/nethserver-blacklist-conf:40]: See the comment below, the URL is not more a mandatory, we use it like a new checkbox in the UI, if the URL is a valid git repository and if the URL exists then we download the list and use this list to the cronjob

Comment (gsanchietti) [reply root/etc/e-smith/events/actions/nethserver-blacklist-conf:40]: Not sure if it's a good choice, I need to test it.
I will do it probably tomorrow :)

**Comment by stephdl** [comment]
obsoleted by https://github.com/NethServer/nethserver-blacklist/pull/38

**Pull request closed without merging**